### PR TITLE
[Feature] Week 4 Features

### PR DESCRIPTION
**Pull request by daekoon**
## Changes
- Added basic algorithm for the algo project

### CODE REVIEW

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,55 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: Where is the validation for the `noOptionsForTooLongOfARequest` test?
I think the test passes because of the empty events being passed?

Seems also the same for the `optionsForNoAttendees` test.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,55 @@
...
+    List<TimeRange> possibleSchedule = new ArrayList<TimeRange>();
+
+    Collections.sort(existingEvents);
+    int curStartTime = 0;
+    TimeRange eventTimeRange;
+    boolean contains;
+
+    for (Event event : existingEvents) {
+      contains = false;
+      for (String participant : request.getAttendees()) {
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:34]: Alternatively you can also use `Collections.disjoint` to do this.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,55 @@
...
+      }
+
+      if (eventTimeRange.start() - curStartTime >= request.getDuration()) {
+        possibleSchedule.add(TimeRange.fromStartDuration(curStartTime, eventTimeRange.start() - curStartTime));
+      }
+      curStartTime = eventTimeRange.end();
+
+    }
+
+    if (curStartTime != TimeRange.END_OF_DAY) {
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:60]: Can the if-statement-block potentially be refactored into a smaller function?

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: Is it necessary to have an extra validation for these two edge cases, when my original algorithm already handles it the way the tests want it to be?

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:60]: Are you referring to the various if blocks that are used to handle each cases? I did not refactor them out to separate functions because they are simple enough. Or are you asking me to refactor it to increase readability?

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Replaced code with Collections.disjoint

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (henryken) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: This is a fair argument assuming that the test cases perfectly cover the possible scenarios.

If you look at those test cases, the `query` method is passed in `NO_EVENTS`.
What if there are some events scheduled and passed to the `query` method?
Does your logic cover that?

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (henryken) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: There are multiple ways things can be refactored:
* To extract the if-condition expression to separate method
* To extract the whole if-block to separate method

It definitely looks simple enough, especially when you are the writer of the code.
But cognitively for other code readers, it is more difficult to comprehend.
It can help to explain your code better, encapsulate logic without revealing the details, easier maintenance, etc.

For this, one suggestion you can give a try for illustration: `addToScheduleIfMeetingFitsAtEndOfDay(...)`

Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:68]: Refactored the if-else statements.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: Added extra tests to show that these cases are covered, even if there are events passed in to queries.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,61 @@
...
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> existingEvents = new ArrayList<Event>(events);
+    List<TimeRange> possibleSchedule = new ArrayList<TimeRange>();
+
+    Collections.sort(existingEvents);
+    int curStartTime = 0;
+    TimeRange eventTimeRange;
+    boolean contains;
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:30]: This is not used anymore

File: walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 5]
```diff
@@ -64,6 +66,12 @@ public void optionsForNoAttendees() {
     Collection<TimeRange> expected = Arrays.asList(TimeRange.WHOLE_DAY);
 
     Assert.assertEquals(expected, actual);
+
+    // If there are other events scheduled in the day
+    actual = query.query(Arrays.asList(Events.events), request);
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:71]: Additional test should be in separate test case.
This ensures that the test is testing one thing at a time.

File: walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 6]
```diff
@@ -76,6 +84,18 @@ public void noOptionsForTooLongOfARequest() {
     Collection<TimeRange> expected = Arrays.asList();
 
     Assert.assertEquals(expected, actual);
+
+    // Case when there are other events scheduled
+    Collection<Event> testCaseWithSchedules = new ArrayList<Event>(Arrays.asList(Events.events));
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:89]: Additional test should be in separate test case.
This ensures that the test is testing one thing at a time.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:30]: This is not used anymore.

File: walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 5 above)
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:71]: Additional test should be in separate test case.
This ensures that the test is testing one thing at a time.

File: walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:89]: Additional test should be in separate test case.
This ensures that the test is testing one thing at a time.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (henryken) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:23]: Thanks for adding the extra test cases.

Although the logic handles it well, I think it can be improved further by doing it as separate validation logic at the beginning.
The current approach needs to iterate all the events unnecessarily for a meeting request that has no attendee or the duration is too long.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,61 @@
...
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> existingEvents = new ArrayList<Event>(events);
```
Comment (henryken) [inline walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:24]: Although the logic handles it well, I think it can be improved further by doing it as separate validation logic at the beginning.
The current approach needs to iterate all the events unnecessarily for a meeting request that has no attendee or the duration is too long.

**Review by henryken — COMMENTED** [review]
Thanks for making the changes!
Just to push you a little bit more, I left some comments.

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Removed unused variable

File: walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 5 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java]: Separated Tests

File: walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java (hunk 6 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:98]: Separated Tests

File: walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (daekoon) [reply walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: Added additional validation logic for edge cases

**Comment by daekoon** [comment]
Made all the necessary changes.
I think my mistake was that I was too focused on this being an algorithm problem - where I usually only have to worry about the worst case runtime. In a real software engineering scenario, average case, best case and edge case runtimes actually matter and I did not take those into consideration.

**Review by henryken — APPROVED** [review]
Thanks for making the changes!
